src/contracts/MslV6.ts

```diff
@@ -356,6 +356,73 @@ export class MslV6 extends BaseContract<typeof multiSourceLoanAbiV6> {
     return lockupTimeSeconds - ellapsedSeconds;
   }
 
+  async refinanceBatch({
+    renegotiationId,
+    refinancings,
+  }: {
+    renegotiationId: bigint;
+    refinancings: {
+      loan: LoanV6;
+      newAprBps: bigint;
+      sources: {
+        source: LoanV6['tranche'][number] & { loanIndex: number };
+        refinancingPrincipal: bigint;
+      }[];
+    }[];
+  }) {
+    // Generate multicall encoded function data for (renegotiation offer, loan) pairs
+    const data = refinancings.map(({ loan, sources, newAprBps }, index) => {
+      const trancheIndex = sources.map(({ source }) => BigInt(source.loanIndex));
+
+      const refinancingPrincipalAmount = sources.reduce(
+        (acc, { refinancingPrincipal }) => acc + refinancingPrincipal,
+        0n,
+      );
+
+      const offer = {
+        renegotiationId: renegotiationId + BigInt(index),
+        loanId: BigInt(Math.max(...loan.tranche.map(({ loanId }) => Number(loanId)))),
```

Comment:
I would move this to a loan helper

Reply:
https://github.com/gondixyz/gondi-js/pull/65/commits/877cd89275c0bdbff12bd5e4756caf73ec9110c4